lib/spack/spack/package_base.py

```diff
@@ -1001,6 +1001,42 @@ def detect_dev_src_change(self) -> bool:
         assert dev_path_var and record, "dev_path variant and record must be present"
         return fsys.recursive_mtime_greater_than(dev_path_var.value, record.installation_time)
 
+    @classmethod
+    def version_or_package_attr(cls, attr, version, default=None):
+        """
+        Get an attribute that could be on the version or package with preference to the version
+        """
+        version_attrs = cls.versions.get(version)
+        if version_attrs and attr in version_attrs:
+            return version_attrs.get(attr)
+        value = getattr(cls, attr, default)
+        if value is None:
+            raise PackageError(f"{attr} attribute not defined on {cls.name}")
+        return value
```

Comment:
This code doesn't do quite what you want -- the value could actually be `None`, in which case you will inappropriately raise. I think you need an otherwise impossible default arg so that a `None` default can be valid as well. There's a couple places in the code where we use `object()` to construct a default

```suggestion
    no_default = object()  # Probably better to move this outside the class to module scope

    @classmethod
    def version_or_package_attr(cls, attr, version, default=no_default):  # self.no_default if you don't move it outside module scope
        """
        Get an attribute that could be on the version or package with preference to the version
        """
        version_attrs = cls.versions.get(version)
        if version_attrs and attr in version_attrs:
            return version_attrs.get(attr)
        if default is no_default and not hasattr(cls, attr):
            raise PackageError(f"{attr} attribute not defined on {cls.name}")
        return getattr(cls, attr, default)
```